File: cppCode/lexer.cpp
#include "lexer.h"
#include <iostream>
#include <regex>

std::vector<Token> lexer(const std::string &input)
{
    std::vector<Token> tokens;
    size_t pos = 0;

    while (pos < input.length())
    {
        char char_at = input[pos];

        if (std::isspace(char_at))
        {
            pos++;
            continue;
        }

        if (std::isalpha(char_at))
        {
            std::string value;
            while (pos < input.length() && std::isalnum(input[pos]))
            {
                value += input[pos++];
            }
            const std::string keywords[] = {"int", "for", "while", "do", "if", "else", "switch", "case", "default", "break"};
            bool isKeyword = false;
            for (const std::string &kw : keywords)
            {
                if (value == kw)
                {
                    std::string upper_kw = kw;
                    for (char &c : upper_kw)
                        c = std::toupper(c);
                    tokens.push_back(Token(upper_kw, value));
                    isKeyword = true;
                    break;
                }
            }
            if (!isKeyword)
                tokens.push_back(Token("IDENTIFIER", value));
            continue;
        }

        if (std::isdigit(char_at))
        {
            std::string value;
            while (pos < input.length() && std::isdigit(input[pos]))
            {
                value += input[pos++];
            }
            tokens.push_back(Token("NUMBER", value));
            continue;
        }

        std::string single_chars = "+-*/%=<>(){};:,";
        const char *types[] = {"PLUS", "MINUS", "MULT", "DIV", "MOD", "EQUALS", "LT", "GT", "LPAREN", "RPAREN", "LBRACE", "RBRACE", "SEMICOLON", "COLON", "COMMA"};
        size_t idx = single_chars.find(char_at);
        if (idx != std::string::npos)
        {
            tokens.push_back(Token(types[idx], std::string(1, char_at)));
            pos++;
            continue;
        }

        if (char_at == '=' && pos + 1 < input.length() && input[pos + 1] == '=')
        {
            tokens.push_back(Token("EQ", "=="));
            pos += 2;
            continue;
        }
        if (char_at == '!' && pos + 1 < input.length() && input[pos + 1] == '=')
        {
            tokens.push_back(Token("NEQ", "!="));
            pos += 2;
            continue;
        }
        if (char_at == '<' && pos + 1 < input.length() && input[pos + 1] == '=')
        {
            tokens.push_back(Token("LTE", "<="));
            pos += 2;
            continue;
        }
        if (char_at == '>' && pos + 1 < input.length() && input[pos + 1] == '=')
        {
            tokens.push_back(Token("GTE", ">="));
            pos += 2;
            continue;
        }

        pos++;
    }

    std::cout << "Output from Lexer stage: ";
    for (const Token &t : tokens)
    {
        std::cout << "(" << t.type << ", " << t.value << ") ";
    }
    std::cout << std::endl;

    return tokens;
}

File: JS/lexer.js
const Token = require('./token');

function lexer(input) {
  const tokens = [];
  let pos = 0;

  while (pos < input.length) {
    let char = input[pos];

    if (/\s/.test(char)) {
      pos++;
      continue;
    }

    if (/[a-zA-Z]/.test(char)) {
      let value = '';
      while (pos < input.length && /[a-zA-Z0-9]/.test(input[pos])) {
        value += input[pos++];
      }
      const keywords = ['int', 'for', 'while', 'do', 'if', 'else', 'switch', 'case', 'default', 'break'];
      tokens.push(new Token(keywords.includes(value) ? value.toUpperCase() : 'IDENTIFIER', value));
      continue;
    }

    if (/[0-9]/.test(char)) {
      let value = '';
      while (pos < input.length && /[0-9]/.test(input[pos])) {
        value += input[pos++];
      }
      tokens.push(new Token('NUMBER', value));
      continue;
    }

    const singleChars = {
      '+': 'PLUS', '-': 'MINUS', '*': 'MULT', '/': 'DIV', '%': 'MOD',
      '=': 'EQUALS', '<': 'LT', '>': 'GT', '(': 'LPAREN', ')': 'RPAREN',
      '{': 'LBRACE', '}': 'RBRACE', ';': 'SEMICOLON', ':': 'COLON', ',': 'COMMA'
    };
    if (singleChars[char]) {
      tokens.push(new Token(singleChars[char], char));
      pos++;
      continue;
    }

    if (char === '=' && input[pos + 1] === '=') {
      tokens.push(new Token('EQ', '=='));
      pos += 2;
      continue;
    }
    if (char === '!' && input[pos + 1] === '=') {
      tokens.push(new Token('NEQ', '!='));
      pos += 2;
      continue;
    }
    if (char === '<' && input[pos + 1] === '=') {
      tokens.push(new Token('LTE', '<='));
      pos += 2;
      continue;
    }
    if (char === '>' && input[pos + 1] === '=') {
      tokens.push(new Token('GTE', '>='));
      pos += 2;
      continue;
    }

    pos++;
  }

  console.log('Output from Lexer stage:', tokens.map(t => `(${t.type}, ${t.value})`));
  return tokens;
}

module.exports = lexer;